source/fuzz/transformation_swap_functions.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 Emiljano Gjiriti
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_swap_functions.h"
+
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationSwapFunctions::TransformationSwapFunctions(
+    protobufs::TransformationSwapFunctions message)
+    : message_(std::move(message)) {}
+
+TransformationSwapFunctions::TransformationSwapFunctions(uint32_t result_id1,
+                                                         uint32_t result_id2) {
+  message_.set_result_id1(result_id1);
+  message_.set_result_id2(result_id2);
+}
+
+bool TransformationSwapFunctions::IsApplicable(
+    opt::IRContext *ir_context,
+    const TransformationContext & /*unused*/) const {
+
+  // A swap transformation is possible if:
+  // - Both the result_ids are valid result_ids of some functions in the module
+  // and
+  // - The result_ids are not equal
+  opt::Function *f1 = ir_context->GetFunction(message_.result_id1());
+  opt::Function *f2 = ir_context->GetFunction(message_.result_id2());
+
+  if (f1 == nullptr || f2 == nullptr) {
+    return false;
+  }
+  assert(f1 != nullptr && "Function 1 doesn't exist");
+  assert(f2 != nullptr && "Function 2 doesn't exist");
```

Comment:
Please remove these asserts since they will always succeed - we already an if statement above that checks for nulls.

Reply:
Ok. Do you think it's better to remove the asserts or the if condition?

Reply:
Asserts is better since the transformation can be supplied with invalid ids in the shrinker (it's one of the fuzzer's modes). You don't need to worry about these details for now, though :).

Reply:
Sorry for the confusion. I meant it's better to remove asserts.